xds/src/main/java/io/grpc/xds/WeightedRoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,471 @@
+                      + ", attrs=" + resolvedAddresses.getAttributes()));
+      return false;
+    }
+    config =
+            (WeightedRoundRobinLoadBalancerConfig) resolvedAddresses.getLoadBalancingPolicyConfig();
+    boolean accepted = super.acceptResolvedAddresses(resolvedAddresses);
+    new UpdateWeightTask().run();
+    afterAcceptAddresses();
+    return accepted;
+  }
+
+  @Override
+  public RoundRobinPicker createReadyPicker(List<Subchannel> activeList, int startIndex) {
+    return new WeightedRoundRobinPicker(activeList, startIndex);
+  }
+
+  private final class UpdateWeightTask implements Runnable {
+    @Override
+    public void run() {
+      if (weightUpdateTimer != null && weightUpdateTimer.isPending()) {
+        return;
+      }
+      if (currentPicker != null && currentPicker instanceof WeightedRoundRobinPicker) {
+        ((WeightedRoundRobinPicker)currentPicker).updateWeight();
+      }
+      weightUpdateTimer = syncContext.schedule(new UpdateWeightTask(),
+              config.weightUpdatePeriodNanos, TimeUnit.NANOSECONDS, timeService);
+    }
+  }
+
+  private void afterAcceptAddresses() {
+    for (Subchannel subchannel : getSubchannels()) {
+      WrrSubchannel weightedSubchannel = (WrrSubchannel) subchannel;
+      weightedSubchannel.setConfig(config);
+      if (config.enableOobLoadReport) {
+        OrcaOobUtil.setListener(weightedSubchannel, weightedSubchannel.oobListener,
+                OrcaOobUtil.OrcaReportingConfig.newBuilder()
+                        .setReportInterval(config.oobReportingPeriodNanos, TimeUnit.NANOSECONDS)
+                        .build());
+      } else {
+        OrcaOobUtil.setListener(weightedSubchannel, null, null);
+      }
+    }
+  }
+
+  @Override
+  public void shutdown() {
+    if (weightUpdateTimer != null) {
+      weightUpdateTimer.cancel();
+    }
+    super.shutdown();
+  }
+
+  private static final class WrrHelper extends ForwardingLoadBalancerHelper {
+    private final Helper delegate;
+    private final TimeProvider timeProvider;
+
+    WrrHelper(Helper helper, TimeProvider timeProvider) {
+      this.delegate = helper;
+      this.timeProvider = timeProvider;
+    }
+
+    @Override
+    protected Helper delegate() {
+      return delegate;
+    }
+
+    @Override
+    public Subchannel createSubchannel(CreateSubchannelArgs args) {
+      return new WrrSubchannel(delegate().createSubchannel(args), timeProvider);
+    }
+  }
+
+  @VisibleForTesting
+  static final class WrrSubchannel extends ForwardingSubchannel {
+    private final Subchannel delegate;
+    private final TimeProvider timeProvider;
+    private final OrcaOobReportListener oobListener = this::onLoadReport;
+    private final OrcaPerRequestReportListener perRpcListener = this::onLoadReport;
+    private volatile long lastUpdated;
+    private volatile long nonEmptySince;
+    private volatile double weight;
+    private volatile WeightedRoundRobinLoadBalancerConfig config;
```

Comment:
oh it appears so.
It can be used by pick() from `SubchannelStateListener` from transport thread, or here in the LB thread in weightUpdateTask or acceptResolvedAddresses, all in sync context.